qiling/debugger/qdb/render/render.py

```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/env python3
+#
+# Cross Platform and Multi Architecture Advanced Binary Emulation Framework
+#
+
+from __future__ import annotations
+from typing import Mapping, Iterable
+import os, copy
+
+from ..context import Context
+from ..const import color
+
+
+def get_terminal_size() -> Iterable:
+    """
+    get terminal window height and width
+    """
+
+    return map(int, os.popen('stty size', 'r').read().split())
+
+"""
+
+    Context Render for rendering UI
+
+"""
+
+COLORS = (color.DARKCYAN, color.BLUE, color.RED, color.YELLOW, color.GREEN, color.PURPLE, color.CYAN, color.WHITE)
+
+class Render(object):
+    """
+    base class for rendering related functions
+    """
+
+    def divider_printer(field_name, ruler="─"):
+        """
+        decorator function for printing divider and field name
+        """
+
+        def decorator(context_dumper):
+            def wrapper(*args, **kwargs):
+                height, width = get_terminal_size()
+                bar = (width - len(field_name)) // 2 - 1
+                print(ruler * bar, field_name, ruler * bar)
+                context_dumper(*args, **kwargs)
+                if "DISASM" in field_name:
+                    print(ruler * width)
+            return wrapper
+        return decorator
+
+    def __init__(self) -> Render:
+        self.regs_a_row = 4
+        self.stack_num = 10
+        self.color = color
+
+    def reg_diff(self, cur_regs, saved_reg_dump):
+        """
+        helper function for highlighting register changed during execution
+        """
+
+        if saved_reg_dump:
+            reg_dump = copy.deepcopy(saved_reg_dump)
+            if getattr(self, "regs_need_swaped", None):
+                reg_dump = self.swap_reg_name(reg_dump)
+
+            return [k for k in cur_regs if cur_regs[k] != reg_dump[k]]
+
+    def render_regs_dump(self, regs, diff_reg=None):
+        """
+        helper function for redering registers dump
+        """
+
+        lines = ""
+        for idx, r in enumerate(regs, 1):
+            line = "{}{}: 0x{{:08x}}  {}\t".format(COLORS[(idx-1) // self.regs_a_row], r, color.END)
+
+            if diff_reg and r in diff_reg:
+                line = f"{color.UNDERLINE}{color.BOLD}{line}"
+
+            if idx % self.regs_a_row == 0 and idx != 32:
+                line += "\n"
+
+            lines += line
+
+        print(lines.format(*regs.values()))
+
+    def render_stack_dump(self, arch_sp: int, pointer_size: int = 4) -> None:
+        """
+        helper function for redering stack dump
+        """
+
+        for idx in range(self.stack_num):
+            addr = arch_sp + idx * pointer_size
+            if (val := self.try_read_pointer(addr)[0]):
```

Comment:
This is C code written in Python syntax..
Avoid assigning variables with values and checking them in a single statement.